### PR TITLE
check API version

### LINKED ISSUE #633: Enforcement of API Protocol versions in Client/Signer/Aggregator
## Issue
In order to help signers and client operators know if they are running a stale version of the API protocol and facilitate the overall upgrade process of the nodes, we need to put in place a verification mechanism. This mechanism must be optional, if no version is provider by the HTTP client, the verification is skipped.

We need to have signers and client stop and display an error message requesting for an update of the node in case of incompatible `API_VERSION` detected between signer & aggregator or client & aggregator. 

For instance, the Aggregator could return a specific HTTP error code and this would trigger an exit from the signer/client. 

Protocol versions are compatible if they share the same minor.

## To do

- [ ] If Aggregator version !== Signer/Client version: display an error "You need to upgrade your node to version X.Y"
- [ ] If the version is not provided, the version is ignored.
- [ ] Add units tests or maybe integration tests
- [ ] Maybe add a dev blog post (if it brings added value to SPOs)

Relates to #565

### PR DESCRIPTION
**Pull request by ghubertpalo**
## Content
This PR adds an API version check when requesting the Aggregator. If the API version does not match, the HTTP call is rejected.

## Pre-submit checklist

- Branch
  - [X] Tests are provided (if possible)
  - [X] Commit sequence broadly makes sense
  - [X] Key commits have useful messages
- PR
  - [X] No clippy warnings in the CI
  - [X] Self-reviewed the diff
  - [X] Useful pull request description
  - [X] Reviewer requested
- Documentation
  - [ ] Update README file (if relevant)
  - [ ] Update documentation website (if relevant)
  - [ ] Add dev blog post (if relevant)

## Comments
None

## Issue(s)

Closes #633

### CODE REVIEW

File: mithril-aggregator/src/http_server/routes/router.rs [hunk 1]
```diff
@@ -24,14 +31,75 @@ pub fn routes(
...
+                .or(signer_routes::routes(dependency_manager.clone()))
+                .or(signatures_routes::routes(dependency_manager.clone()))
+                .or(epoch_routes::routes(dependency_manager))
+                .with(cors),
+        )
+        .recover(handle_custom)
+        .with(warp::reply::with::headers(headers))
+}
+
+/// API Version verification
```
Comment (ghubertpalo) [inline mithril-aggregator/src/http_server/routes/router.rs:49]: change the comment according to what the function does

File: mithril-aggregator/src/http_server/routes/router.rs [hunk 2]
```diff
@@ -24,14 +31,75 @@ pub fn routes(
...
+            match maybe_header {
+                None => Ok(()),
+                Some(version) if version == MITHRIL_API_VERSION => Ok(()),
+                Some(_version) => Err(warp::reject::custom(VersionMismatchError)),
+            }
+        })
+        .untuple_one()
+}
+
+pub async fn handle_custom(reject: Rejection) -> Result<impl Reply, Rejection> {
```
Comment (ghubertpalo) [inline mithril-aggregator/src/http_server/routes/router.rs:62]: handle custom rejection codes

Comment (ghubertpalo) [inline mithril-aggregator/src/http_server/routes/router.rs:62]: not need to be public

**Comment by ghubertpalo** [comment]
Catch the HTTP error in the signer and the client to display an according error message.

**Comment by jpraynaud** [comment]
We probably need to update the `OpenAPI` specification to add the new error code.

Also I'm not sure that we should use [`412`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/412) HTTP code. Maybe [`426`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/426) is more appropriate. WDYT?

**Review by jpraynaud — APPROVED** [review]
LGTM :+1:

**Pull request merged**